lockfile/src/cyclonedx.rs

```diff
@@ -0,0 +1,313 @@
+
+    fn version(&self) -> &str {
+        &self.version
+    }
+
+    fn scope(&self) -> Option<&str> {
+        self.scope.as_deref()
+    }
+
+    fn purl(&self) -> Option<&str> {
+        self.purl.as_deref()
+    }
+
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized,
+    {
+        self.components.as_ref().map(|comps| comps.components.as_slice())
+    }
+}
+
+// Represents a single JSON component.
+#[derive(Clone, Debug, Deserialize)]
+struct JsonComponent {
+    #[serde(rename = "type")]
+    component_type: String,
+    name: String,
+    version: String,
+    scope: Option<String>,
+    purl: Option<String>,
+    components: Option<Vec<JsonComponent>>,
+}
+
+impl Component for JsonComponent {
+    fn component_type(&self) -> &str {
+        &self.component_type
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn version(&self) -> &str {
+        &self.version
+    }
+
+    fn scope(&self) -> Option<&str> {
+        self.scope.as_deref()
+    }
+
+    fn purl(&self) -> Option<&str> {
+        self.purl.as_deref()
+    }
+
+    fn components(&self) -> Option<&[Self]>
+    where
+        Self: Sized,
+    {
+        self.components.as_deref()
+    }
+}
+
+// Filter components based on their type and scope.
+fn filter_components<T: Component>(components: &[T]) -> impl Iterator<Item = &'_ T> {
+    components
+        .iter()
+        .filter(|&comp| {
+            let type_check = comp.component_type() == "application"
+                || comp.component_type() == "framework"
+                || comp.component_type() == "library";
+
+            // Check if the scope is "required" or not specified (required)
+            let scope_check = match comp.scope() {
+                Some(scope) => scope == "required",
+                None => true,
+            };
+
+            type_check && scope_check
+        })
+        .flat_map(|comp| {
+            let nested_iter = match comp.components() {
+                Some(nested) => filter_components(nested).collect::<Vec<_>>(),
+                None => Vec::new(),
+            };
+            std::iter::once(comp).chain(nested_iter.into_iter())
+        })
+}
+
+// Convert a component's Package URL (PURL) into a Package object.
+fn from_purl<T: Component>(component: &T) -> anyhow::Result<Package> {
+    let purl_str = component
+        .purl()
+        .ok_or_else(|| anyhow!("Missing purl for {}:{}", component.name(), component.version()))?;
+    let purl = GenericPurl::<String>::from_str(purl_str)?;
+    let package_type = PackageType::from_str(purl.package_type()).map_err(|_| UnknownEcosystem)?;
+
+    // Determine the package name based on its type and potential namespace.
+    let name = match (package_type, purl.namespace()) {
+        (PackageType::Maven, Some(ns)) => format!("{}:{}", ns, purl.name()),
+        (PackageType::Npm | PackageType::Golang, Some(ns)) => format!("{}/{}", ns, purl.name()),
+        _ => purl.name().into(),
+    };
```

Comment:
This looks **very** familiar. Is it possible we did this in the sbom parser already?

If so, isn't it maybe worth it to create an abstraction for this?